File: leetcode/medium/Q146_LRUCache.py
class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.dic = {}
        self.recent = []
    def get(self, key: int) -> int:
        if key in self.recent:
            self.recent.remove(key)
            self.recent.append(key)
            return self.dic[key]
        return -1
    def put(self, key: int, value: int) -> None:
        if len(self.recent) == self.capacity:
            if key not in self.recent:
                self.recent.pop(0)
        if key in self.recent:
            self.recent.remove(key)
        self.recent.append(key)
        self.dic[key] = value

File: src/medium/Q146_LRUCache.java
package medium;

import java.util.*;

public class Q146_LRUCache {
    class LRUCache {
        private Map<Integer, Integer> map;
        private LinkedHashSet<Integer> cache;
        private int capacity;
        public LRUCache(int capacity) {
            this.cache = new LinkedHashSet<>();
            this.map = new HashMap<>();

            this.capacity = capacity;
        }

        public int get(int key) {
            if(map.containsKey(key)){
                cache.remove(key);
                cache.add(key);
                return map.get(key);
            }
            return -1;
        }

        public void put(int key, int value) {
            cache.remove(key);
            cache.add(key);
            if(cache.size() > capacity){
                for(int c : cache){
                    cache.remove(c);
                    map.remove(c);
                    break;
                }
            }
            map.put(key, value);
        }
    }
}
